### PR TITLE
Add --strip that removes debug info

### PR DESCRIPTION
**Pull request by kripken**
This is sort of like `--strip` on a native binary. The more specific use case for us is e.g. you link with a library that has `-g` in its CFLAGS, but you don't want debug info in your final executable (I hit this with poppler now). We can make emcc pass this to binaryen if emcc is not building an output with intended debug info.

This also raised the question of the current wasm-opt `-g` flag. Without that flag, we don't emit the names section, so we strip it by default basically. That's the same behavior as `asm2wasm`, but @dschuff made the point that the former is a compiler, so it makes sense, while `wasm-opt` should probably pass through a names section if one already exists? If we don't want to make that change, then perhaps it makes sense to run `--strip` automatically if `-g` is not found? But then that seems a little odd, so the former option may be better.

### CODE REVIEW

File: src/passes/Strip.cpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    auto& sections = module->userSections;
+    sections.erase(
+      std::remove_if(
+        sections.begin(),
+        sections.end(),
+        [&](const UserSection& curr) {
+          return curr.name == BinaryConsts::UserSections::Name ||
+                 curr.name == BinaryConsts::UserSections::SourceMapUrl ||
+                 curr.name.find(".debug") == 0 ||
+                 curr.name == "reloc..debug_info" ||
```
Comment (yurydelendik) [inline src/passes/Strip.cpp]: Other debug relocation sections maybe present, not only "reloc..debug_info" or "reloc..debug_line", if it is possible use `curr.name.find("reloc..debug") == 0` here instead.

**Comment by kripken** [comment]
Thanks @yurydelendik!

Hmm, not sure what to do with the linking section. It's needed for more than debugging so I don't think we can remove it here. However, yeah, that sounds bad that it can contain references to debug sections that we would be removing here. @sbc100 , what do you think?

**Comment by dschuff** [comment]
Yeah I think we may want to eventually add support for keeping reloc info up to date (without it we can't optimize object files), but we don't have it now, so it will definitely break. So always stripping the static linking sections makes sense. Or maybe we should even just refuse to process object files at all; object files aren't runnable until they are linked, and we will break them if we modify them (and since Binaryen's encode/decode isn't guaranteed to roundtrip we may always modify them), so I don't think running wasm-opt on an object file could ever work right now?

**Comment by yurydelendik** [comment]
>  I'm pretty sure the generated debug info will always be broken because it refers to binary offsets in the code, no?

Correct. If code or data sections are modified in any way, the debug sections will become invalid.

File: src/passes/Strip.cpp (hunk 1 above)
Comment (kripken) [reply src/passes/Strip.cpp]: Thanks, fixing that way.

**Comment by kripken** [comment]
Addressed feedback.

I'll open a separate PR to show a warning or an error if the linking section is present.

**Comment by sbc100** [comment]
In general this information should already be stripped by `lld` if it is run with the corresponding flags:

```
  --strip-all            Strip all symbols
  --strip-debug          Strip debugging information
```

No harm in adding these flang to binaryen too I suppose, but we probably want to put them on lld to save time there too if we don't want this info.

**Comment by kripken** [comment]
@sbc100 thanks, good to know. So we almost don't need this then, but it's still useful for non-lld paths (asm2wasm, other compilers perhaps?).

**Pull request merged**